README.md

```diff
@@ -144,7 +144,11 @@ For "static" feature flags we set the flag values in env vars.
 To add a new feature flag, set a `"#{flag_name.upcase}_FEATURE_FLAG"` env var with value `"ENABLED"` in all environments where you want the flag enabled.
 Then add `flag_name` to the list of flags  in `app/lib/feature_flags.rb`.
 
-To use the feature flag in your code, just call `FeatureFlags.enabled?(:flag_name)`.
+When adding a `flag_name` to the list of static flags, you will need to decide if this is a `"global"` flag i.e. always taken from the env var and not the session, or a `"session"` flag i.e. taken from the `session_data` of the check. 
+
+We introduced this as a way of making our feature flags 'backwards compatible' - if a user check is underway during the switch-on of a flag, their user journey will not be affected and they can continue through their check with the flag value that was set at the beginning of their journey.
+
+To use the feature flag in your code, just call `FeatureFlags.enabled?(:flag_name)`. There is an optional `session_data` argument which you must use if you wish to derive the flag value from the `session_data`. Usage is `FeatureFlags.enabled?(:flag_name, session_data)`.
```

Comment:
If I've understood the code correctly (adding the flag value to the session data is only done for session flags, you can look for a global flag on session data first you just won't find it) I wonder if it would be easier to comprehend if we said that `session_data` arg wasn't optional at all?

My concern is that someone reading the code might think that the passing of `session_data` to the function call _does_ something and thus is a bit of a trap (and obviously if you don't pass it to one that's a session flag you don't get the correct behaviour, which again could be a trickier one to debug).

Reply:
Yes I was also not sure on the optional arg and agree with you - the reason it is optional currently, is that when the application loads, we have flags set on the start page that need to determine what to show, before a check has been created. So we can't use a required check argument with `def enabled?`

The other option that @patrick-laa and I discussed was creating two methods instead, something like `def globally_enabled?` and `def session_enabled?` which would have relevant code and required arguments, and you'd have to pick which one you're using depending on the use case. What do you think? Happy to look at implementing that if you think it is better but we wondered if it might be overkill. You're right though, it would be easy not to pass in session_data when it is needed.

Reply:
Ooh yes interesting, I hadn't thought of that edge case! I wonder if we could keep it technically optional but conventionally required ie we say "if you have a session with a check in, pass it, always"?

The other option, with different lookup methods, sounds appealing but does take us further from the simple, agnostic lookup. Maybe that's a good thing? Being explicit at lookup time. Might make the behaviour easier to understand, as it's closer to where it's used.

This is absolutely not my ultimate call though, so I'll just leave it at highlighting the risks I see and my thoughts from looking at the code in isolation, and totally trust you folk to implement what's best given that info 👍

Reply:
How about there are 3 arguments to enabled - `flag_name` (mandatory), `session_data` (can be nil), and a keyword argument `i_understand_that_i_should_always_be_providing_session_data_except_in_very_particular_circumstances` which defaults to false. So this is valid:

```ruby
FeatureFlags.enabled?(:foo, session_data)
```

And so is this:

```ruby
FeatureFlags.enabled?(:foo, nil, i_understand_that_i_should_always_be_providing_session_data_except_in_very_particular_circumstances: true)
```

But this will raise an exception:

```ruby
FeatureFlags.enabled?(:foo, nil)
```

That way it's very hard for someone to accidentally do the wrong thing and not notice.

Reply:
@patrick-laa I quite like this idea of passing in the additional argument. It makes it very hard to get it wrong without realising. Will work with that and see what it looks like.

Reply:
I've implemented the above now and updated the README so I think this is ready for another review if possible!

Reply:
Nice approach